File: LeetCode/295.Find_Median_from_Data_Stream/Mario/295.Find_Median_from_Data_Stream.java
class MedianFinder {

    private PriorityQueue<Integer> minHeap;
    private PriorityQueue<Integer> maxHeap;

    public MedianFinder() {
        minHeap = new PriorityQueue<>((a, b) -> a - b);
        maxHeap = new PriorityQueue<>((a, b) -> b - a);
    }

    public void addNum(int num) {
        maxHeap.add(num);
    }

    public double findMedian() {
        while (!maxHeap.isEmpty() && !minHeap.isEmpty() && maxHeap.peek() > minHeap.peek()) {
            minHeap.add(maxHeap.poll());
        }

        while (minHeap.size() > maxHeap.size()) {
            maxHeap.add(minHeap.poll());
        }

        while (maxHeap.size() > minHeap.size()) {
            minHeap.add(maxHeap.poll());
        }

        if (maxHeap.size() == minHeap.size()) {
            return (maxHeap.peek() + minHeap.peek()) / 2.0;
        } else {
            return minHeap.peek();
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: LeetCode/295.Find_Median_from_Data_Stream/295.Find_Median_from_Data_Stream.py
#!/usr/bin/env python3
import heapq


class MedianFinder:

    # heap, max heap, min heap
    # arr = [2,3,4]
    # low = [-2] # max heap (negative)
    # high = [3, 4] # min heap

    def __init__(self):
        self.low, self.high = [], []

    def addNum(self, num: int) -> None:
        heapq.heappush(self.low, -num)
        heapq.heappush(self.high, -heapq.heappop(self.low))

        # balance, make sure len(self.high) == (len(self.low) or len(self.low) + 1)
        if len(self.high) > len(self.low) + 1:
            heapq.heappush(self.low, -heapq.heappop(self.high))

    def findMedian(self) -> float:
        n = len(self.low) + len(self.high)
        if n % 2 == 1:
            return float(self.high[0])
        else:
            return float(-self.low[0] + self.high[0]) / 2.0


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
